cirq/circuits/circuit.py

```diff
@@ -1258,6 +1258,109 @@ def _json_dict_(self):
     def _from_json_dict_(cls, moments, device, **kwargs):
         return cls(moments, strategy=InsertStrategy.EARLIEST, device=device)
 
+    def tetris_concat(
+        *circuits: 'cirq.AbstractCircuit', stop_at_first_alignment: bool = False
+    ) -> 'cirq.Circuit':
+        """Concatenates circuits while overlapping them if possible.
+
+        Starts with the first circuit (index 0), then iterates over the other
+        circuits while folding them in. To fold two circuits together, they
+        are placed one after the other and then moved inward until any of:
+
+            a) Just before their operations would collide.
+            b) stop_at_first_alignment==True and their ends (or starts) align
+                for the first time.
+            c) Their ends (or starts) align for the last time.
+
+        Beware that this method is *not* associative. For example:
+
+            >>> a, b = cirq.LineQubit.range(2)
+            >>> A = cirq.Circuit(cirq.H(a))
+            >>> B = cirq.Circuit(cirq.H(b))
+            >>> f = cirq.Circuit.tetris_concat
+            >>> f(f(A, B), A) == f(A, f(B, A))
+            False
+            >>> len(f(f(f(A, B), A), B)) == len(f(f(A, f(B, A)), B))
+            False
+
+        Args:
+            circuits: The circuits to concatenate.
+            stop_at_first_alignment: Defaults to false. When true, the circuits
+                are never overlapped more than needed to align their starts or
+                to align their ends. When false, the smaller circuit can be
+                pushed deeper into the larger circuit, past the first time their
+                starts or ends align, until the second time their starts or ends
+                align.
+
+        Returns:
+            The concatenated and overlapped circuit.
+        """
+        if len(circuits) == 0:
+            return Circuit()
+        n_acc = len(circuits[0])
+
+        # Allocate a buffer large enough to append and prepend all the circuits.
+        pad_len = sum(len(c) for c in circuits) - n_acc
+        buffer = np.zeros(shape=pad_len * 2 + n_acc, dtype=np.object)
```

Comment:
The numpy array is intentional, for performance. This is both because a) it has better constant factors and b) because adding together lists can have linear complexity per addition whereas by using the buffer I know for sure I'm not introducing unexpected quadratic behavior.

The constructor takes an iterable of anything that's moments or operations, and I pass it a slice of something that's moments or operations. I agree that when numpy type annotations become possible there will need to be an explicit cast.